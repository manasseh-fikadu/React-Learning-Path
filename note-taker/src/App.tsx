import React from "react";
import "./App.css";
import NoteList from "./components/NotesList";
import NoteForm from "./components/NoteForm";

function App() {
  return (
    // Main container with background color and flex layout
    <div className="bg-gray-900 min-h-screen flex flex-col items-center justify-center text-white">
      {/* Page title */}
      <h1 className="text-4xl font-semibold mb-8">Note Manager</h1>
      {/* Container for the note form */}
      <div className="bg-gray-800 p-6 rounded-lg shadow-md w-96">
        <NoteForm /> {/* Render the NoteForm component */}
      </div>
      {/* Container for the note list */}
      <div className="mt-8 w-96">
        <NoteList /> {/* Render the NoteList component */}
      </div>
    </div>
  );
}

export default App;
