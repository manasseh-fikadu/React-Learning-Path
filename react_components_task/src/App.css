.App {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  background-color: #f0f0f0;
}

.card {
  border: 1px solid #ccc;
  padding: 20px;
  width: 300px;
  text-align: center;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
  border-radius: 8px;
  background-color: #fff;
  transition: transform 0.2s ease;
}

.card:hover {
  transform: translateY(-5px);
}

.avatar {
  width: 100px;
  height: 100px;
  border-radius: 50%;
  overflow: hidden;
  margin: 0 auto;
}

.avatar img {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.info {
  margin-top: 20px;
  color: #444;
}

.info h2 {
  font-size: 1.5rem;
  margin-bottom: 5px;
}

.info p {
  font-size: 1rem;
  line-height: 1.5;
}

a {
  display: inline-block;
  padding: 5px 10px;
  background-color: #007BFF;
  color: #fff;
  text-decoration: none;
  border-radius: 4px;
  margin-top: 10px;
  transition: background-color 0.2s ease;
}

a:hover {
  background-color: #0056b3;
}
