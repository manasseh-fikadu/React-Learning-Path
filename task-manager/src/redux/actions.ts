// Action types
export const ADD_TASK = "ADD_TASK";
export const UPDATE_TASK = "UPDATE_TASK";
export const REMOVE_TASK = "REMOVE_TASK";
export const TOGGLE_TASK_STATUS = "TOGGLE_TASK_STATUS";
export const SET_FILTER = "SET_FILTER";

// Define action interfaces
interface AddTaskAction {
  type: typeof ADD_TASK;
  payload: { text: string; completed: boolean };
}

interface UpdateTaskAction {
  type: typeof UPDATE_TASK;
  payload: { index: number; task: string };
}

interface ToggleTaskStatusAction {
  type: typeof TOGGLE_TASK_STATUS;
  payload: number; // Index of the task to toggle
}

interface RemoveTaskAction {
  type: typeof REMOVE_TASK;
  payload: number; // Index of the task to remove
}

interface SetFilterAction {
  type: typeof SET_FILTER;
  payload: string; // "all", "completed", or "incomplete"
}

// Combine all action types into a union type
export type TaskActionTypes =
  | AddTaskAction
  | UpdateTaskAction
  | RemoveTaskAction
  | ToggleTaskStatusAction
  | SetFilterAction;

// Action creators
export const addTask = (task: {
  text: string;
  completed: boolean;
}): TaskActionTypes => ({
  type: ADD_TASK,
  payload: task,
});

export const updateTask = (index: number, task: string): TaskActionTypes => ({
  type: UPDATE_TASK,
  payload: { index, task },
});

export const removeTask = (index: number): TaskActionTypes => ({
  type: REMOVE_TASK,
  payload: index,
});

export const toggleTaskStatus = (index: number): TaskActionTypes => ({
  type: TOGGLE_TASK_STATUS,
  payload: index,
});

export const setFilter = (filter: string): TaskActionTypes => ({
  type: SET_FILTER,
  payload: filter,
});
