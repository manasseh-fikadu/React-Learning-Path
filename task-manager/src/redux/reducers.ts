import {
  ADD_TASK,
  // UPDATE_TASK,
  REMOVE_TASK,
  TaskActionTypes,
  TOGGLE_TASK_STATUS,
  SET_FILTER,
} from "./actions";

// Define the structure of a single task
interface Task {
  text: string;
  completed: boolean;
}

// Define the structure of the task state
interface TaskState {
  tasks: Task[];
  filter: string; // Add filter property to the state
}

// Initial state with an empty task array and a default filter value
const initialState: TaskState = {
  tasks: [],
  filter: "all", // Default filter value
};

// Task reducer
const taskReducer = (
  state = initialState,
  action: TaskActionTypes
): TaskState => {
  switch (action.type) {
    case ADD_TASK:
      return {
        ...state,
        tasks: [...state.tasks, action.payload],
      };
    // case UPDATE_TASK:
    //   return {
    //     ...state,
    //     tasks: state.tasks.map((task, index) =>
    //       index === action.payload.index ? action.payload.task : task
    //     ),
    //   };
    case REMOVE_TASK:
      return {
        ...state,
        tasks: state.tasks.filter((_, index) => index !== action.payload),
      };
    case TOGGLE_TASK_STATUS:
      return {
        ...state,
        tasks: state.tasks.map((task, index) =>
          index === action.payload
            ? { ...task, completed: !task.completed }
            : task
        ),
      };
    case SET_FILTER:
      return {
        ...state,
        filter: action.payload,
      };
    default:
      return state;
  }
};

export default taskReducer;
