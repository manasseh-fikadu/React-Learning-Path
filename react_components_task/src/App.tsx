import React from "react";
import "./App.css";
import Card from "./components/Card";
import { userProfile } from "./data";

function App() {
  return (
    <div className="App">
      <Card profile={userProfile} />
    </div>
  );
}

export default App;
