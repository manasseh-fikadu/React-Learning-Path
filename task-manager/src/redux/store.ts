import { createStore } from "redux";
import taskReducer from "./reducers";

// Define the structure of a single task
export interface Task {
  text: string;
  completed: boolean;
}

// Define the structure of the root state for the Redux store
export interface RootState {
  tasks: Task[]; // Array of tasks
  filter: string; // Current filter for task list
}

// Create the Redux store using the taskReducer and Redux DevTools extension
const store = createStore(
  taskReducer,
  (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
    (window as any).__REDUX_DEVTOOLS_EXTENSION__()
);

export default store;
