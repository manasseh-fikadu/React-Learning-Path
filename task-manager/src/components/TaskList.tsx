import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../redux/store";
import { toggleTaskStatus, removeTask, setFilter } from "../redux/actions";

const TaskList: React.FC = () => {
  const dispatch = useDispatch();
  const tasks = useSelector((state: RootState) => state.tasks);
  const filter = useSelector((state: RootState) => state.filter);
  const [error, setError] = useState(""); // Add error state


  const handleToggleTaskStatus = (index: number) => {
    if (index >= 0 && index < tasks.length) {
      dispatch(toggleTaskStatus(index));
      setError(""); // Clear error on success
    } else {
      setError("Task not found");
    }
  };

  const handleRemoveTask = (index: number) => {
    if (index >= 0 && index < tasks.length) {
      dispatch(removeTask(index));
      setError(""); // Clear error on success
    } else {
      setError("Task not found");
    }
  };

  // Handle setting the filter
  const handleSetFilter = (newFilter: string) => {
    dispatch(setFilter(newFilter));
  };

  // Filter tasks based on the selected filter
  const filteredTasks = filter === "all" ? tasks : tasks.filter((task) => {
    return filter === "completed" ? task.completed : !task.completed;
  });

  return (
    // Container for the task list with shadow and padding
    <div className="bg-white p-6 rounded-lg shadow-md">
      {/* Heading */}
      <h2 className="text-2xl font-semibold mb-4">Task List</h2>
      {/* Filter buttons */}
      <div className="flex space-x-4 mb-4">
        <button
          onClick={() => handleSetFilter("all")}
          className={`px-4 py-2 rounded-lg ${
            filter === "all"
              ? "bg-blue-500 text-white"
              : "bg-gray-200 text-gray-800"
          } focus:outline-none`}
        >
          All
        </button>
        <button
          onClick={() => handleSetFilter("completed")}
          className={`px-4 py-2 rounded-lg ${
            filter === "completed"
              ? "bg-blue-500 text-white"
              : "bg-gray-200 text-gray-800"
          } focus:outline-none`}
        >
          Completed
        </button>
        <button
          onClick={() => handleSetFilter("incomplete")}
          className={`px-4 py-2 rounded-lg ${
            filter === "incomplete"
              ? "bg-blue-500 text-white"
              : "bg-gray-200 text-gray-800"
          } focus:outline-none`}
        >
          Incomplete
        </button>
      </div>
      {/* Scrollable table for task display */}
      {error && <p className="text-red-500 mb-4">{error}</p>} {/* Display error message */}
      <div className="overflow-x-auto">
        <table className="w-full border-collapse">
          <thead>
            <tr>
              <th className="border p-2 text-left">Status</th>
              <th className="border p-2 text-left">Task</th>
              <th className="border p-2 text-left">Actions</th>
            </tr>
          </thead>
          <tbody>
            {filteredTasks.map((task, index) => (
              <tr key={index} className="border">
                <td className="border p-2">
                  {/* Status button */}
                  <button
                    onClick={() => handleToggleTaskStatus(index)}
                    data-testid="statusButton"
                    className={`rounded-lg p-1 ${
                      task.completed ? "text-white" : "text-white"
                    }`}
                  >
                    {task.completed ? "✅" : "❌"}
                  </button>
                </td>
                {/* Task text */}
                <td className="border p-2">{task.text}</td>
                {/* Remove button */}
                <td className="border p-2">
                  <button
                    onClick={() => handleRemoveTask(index)}
                    className="px-2 py-1 bg-red-500 hover:bg-red-600 text-white rounded-lg focus:outline-none"
                  >
                    Remove
                  </button>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};

export default TaskList;
