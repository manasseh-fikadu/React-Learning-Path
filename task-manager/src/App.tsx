import React from "react";
import TaskForm from "./components/TaskForm";
import TaskList from "./components/TaskList";

const App: React.FC = () => {
  return (
    // Outer container with background, min-height, and flex alignment
    <div className="bg-gray-100 min-h-screen flex justify-center items-center py-10">
      {/* Inner container with responsive width */}
      <div className="w-full sm:w-4/5 md:w-3/5 lg:w-2/5">
        {/* Card-like container */}
        <div className="bg-white p-6 rounded-lg shadow-md">
          {/* Heading */}
          <h1 className="text-3xl font-bold mb-6 text-center">Task Manager</h1>
          {/* TaskForm component */}
          <TaskForm />
        </div>
        {/* Margin for separation */}
        <div className="mt-6">
          {/* TaskList component */}
          <TaskList />
        </div>
      </div>
    </div>
  );
};

export default App;
