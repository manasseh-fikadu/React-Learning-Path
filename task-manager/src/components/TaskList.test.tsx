import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import App from "../App";
import TaskList from "./TaskList";
import { addTask } from "../redux/actions";
import { toggleTaskStatus, removeTask, setFilter } from "../redux/actions";

const mockStore = configureStore([]);

describe("App component", () => {
  let store: any;
  let component: any;

  beforeEach(() => {
    store = mockStore({
      tasks: [],
      filter: "All",
    });
    component = render(
      <Provider store={store}>
        <App />
      </Provider>
    );
  });

  it("should render with given state from Redux store", () => {
    expect(component).toMatchSnapshot();
  });

  it("should set the filter when a filter button is clicked", () => {
    const { getByText } = component;

    // Click the 'Completed' filter button
    const filterButton = getByText("Completed");
    fireEvent.click(filterButton);

    // Verify the dispatched action
    const actions = store.getActions();
    expect(actions).toEqual([setFilter("completed")]);
  });
});
