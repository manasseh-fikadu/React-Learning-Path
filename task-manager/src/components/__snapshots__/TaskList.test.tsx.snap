// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render with given state from Redux store 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bg-gray-100 min-h-screen flex justify-center items-center py-10"
      >
        <div
          class="w-full sm:w-4/5 md:w-3/5 lg:w-2/5"
        >
          <div
            class="bg-white p-6 rounded-lg shadow-md"
          >
            <h1
              class="text-3xl font-bold mb-6 text-center"
            >
              Task Manager
            </h1>
            <div
              class="bg-white p-6 rounded-lg shadow-md"
            >
              <h2
                class="text-2xl font-semibold mb-4"
              >
                Add New Task
              </h2>
               
              <form
                id="task-form"
              >
                <div
                  class="flex items-center space-x-4"
                >
                  <input
                    class="flex-grow bg-gray-100 border border-gray-300 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-400"
                    placeholder="Enter task"
                    type="text"
                    value=""
                  />
                  <button
                    class="px-6 py-2 bg-blue-500 hover:bg-blue-600 text-white rounded-lg focus:outline-none"
                  >
                    Add Task
                  </button>
                </div>
              </form>
            </div>
          </div>
          <div
            class="mt-6"
          >
            <div
              class="bg-white p-6 rounded-lg shadow-md"
            >
              <h2
                class="text-2xl font-semibold mb-4"
              >
                Task List
              </h2>
              <div
                class="flex space-x-4 mb-4"
              >
                <button
                  class="px-4 py-2 rounded-lg bg-gray-200 text-gray-800 focus:outline-none"
                >
                  All
                </button>
                <button
                  class="px-4 py-2 rounded-lg bg-gray-200 text-gray-800 focus:outline-none"
                >
                  Completed
                </button>
                <button
                  class="px-4 py-2 rounded-lg bg-gray-200 text-gray-800 focus:outline-none"
                >
                  Incomplete
                </button>
              </div>
               
              <div
                class="overflow-x-auto"
              >
                <table
                  class="w-full border-collapse"
                >
                  <thead>
                    <tr>
                      <th
                        class="border p-2 text-left"
                      >
                        Status
                      </th>
                      <th
                        class="border p-2 text-left"
                      >
                        Task
                      </th>
                      <th
                        class="border p-2 text-left"
                      >
                        Actions
                      </th>
                    </tr>
                  </thead>
                  <tbody />
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-gray-100 min-h-screen flex justify-center items-center py-10"
    >
      <div
        class="w-full sm:w-4/5 md:w-3/5 lg:w-2/5"
      >
        <div
          class="bg-white p-6 rounded-lg shadow-md"
        >
          <h1
            class="text-3xl font-bold mb-6 text-center"
          >
            Task Manager
          </h1>
          <div
            class="bg-white p-6 rounded-lg shadow-md"
          >
            <h2
              class="text-2xl font-semibold mb-4"
            >
              Add New Task
            </h2>
             
            <form
              id="task-form"
            >
              <div
                class="flex items-center space-x-4"
              >
                <input
                  class="flex-grow bg-gray-100 border border-gray-300 rounded-lg px-4 py-2 focus:outline-none focus:border-blue-400"
                  placeholder="Enter task"
                  type="text"
                  value=""
                />
                <button
                  class="px-6 py-2 bg-blue-500 hover:bg-blue-600 text-white rounded-lg focus:outline-none"
                >
                  Add Task
                </button>
              </div>
            </form>
          </div>
        </div>
        <div
          class="mt-6"
        >
          <div
            class="bg-white p-6 rounded-lg shadow-md"
          >
            <h2
              class="text-2xl font-semibold mb-4"
            >
              Task List
            </h2>
            <div
              class="flex space-x-4 mb-4"
            >
              <button
                class="px-4 py-2 rounded-lg bg-gray-200 text-gray-800 focus:outline-none"
              >
                All
              </button>
              <button
                class="px-4 py-2 rounded-lg bg-gray-200 text-gray-800 focus:outline-none"
              >
                Completed
              </button>
              <button
                class="px-4 py-2 rounded-lg bg-gray-200 text-gray-800 focus:outline-none"
              >
                Incomplete
              </button>
            </div>
             
            <div
              class="overflow-x-auto"
            >
              <table
                class="w-full border-collapse"
              >
                <thead>
                  <tr>
                    <th
                      class="border p-2 text-left"
                    >
                      Status
                    </th>
                    <th
                      class="border p-2 text-left"
                    >
                      Task
                    </th>
                    <th
                      class="border p-2 text-left"
                    >
                      Actions
                    </th>
                  </tr>
                </thead>
                <tbody />
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
